contrib/babelfishpg_tsql/src/pl_exec-2.c

```diff
@@ -236,6 +242,126 @@ exec_tsql_stmt(PLtsql_execstate *estate, PLtsql_stmt *stmt, PLtsql_stmt *save_es
 	return rc;
 }
 
+static int
+exec_stmt_kill(PLtsql_execstate *estate, PLtsql_stmt_kill *stmt)
+{
+	PGPROC *proc;    	
+	Oid	    sysadmin_oid = get_role_oid("sysadmin", false);  // should really use BABELFISH_SYSADMIN in tds_int.h
+	int	    spid = -1;
+	char   *strtol_endptr;
+
+	if (pltsql_explain_only)
+	{
+		StringInfoData query;
+
+		initStringInfo(&query);
+		appendStringInfo(&query, "KILL ");
+		appendStringInfoString(&query, stmt->spid);
+		append_explain_info(NULL, query.data);
+		return PLTSQL_RC_OK;
+	}
+
+	// Do not allow to run KILL inside a transaction
+	if (IsTransactionBlockActive())
+	{
+		ereport(ERROR,
+			(errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
+				errmsg("%s command cannot be used inside user transactions.", "KILL")));
+	}
+
+	// Require that the user has 'sysadmin' role
+	if (!has_privs_of_role(GetSessionUserId(), sysadmin_oid)) 
+		{	       
+		ereport(ERROR,
+			(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+				errmsg("User does not have permission to use the KILL statement")));
+	}
+
+	// Convert SPID string to int
+	// Although the T-SQL grammar only accepts a numeric value, and it must always be present, play it safe
+	Assert(stmt->spid != NULL);     
+	Assert(strlen(stmt->spid) > 0);
+
+	spid = (int) strtol(stmt->spid, &strtol_endptr, 10);
```

Comment:
Must convert in parser and store the int/long value

Reply:
done